src/Components/Web/src/Virtualization/VirtualizeDeferred.cs

```diff
@@ -0,0 +1,138 @@
+    public sealed class VirtualizeDeferred<TItem> : VirtualizeBase<TItem>
+    {
+        private enum FetchState
+        {
+            Idle,
+            Waiting,
+            Busy
+        }
+
+        private readonly ConcurrentQueue<TItem> _loadedItems = new ConcurrentQueue<TItem>();
+
+        private FetchState _fetchState;
+
+        private int _itemCount;
+
+        private Task<ItemsProviderResult<TItem>>? _fetchTask;
+
+        private CancellationTokenSource? _fetchTcs;
+
+        /// <summary>
+        /// Gets or sets the item template for the list.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<TItem> Item { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the template for items that have not yet been loaded in memory.
+        /// </summary>
+        [Parameter]
+        public RenderFragment<int> Placeholder { get; set; } = default!;
+
+        /// <summary>
+        /// Gets or sets the function providing items to the list.
+        /// </summary>
+        [Parameter]
+        public ItemsProviderDelegate<TItem> ItemsProvider { get; set; } = default!;
+
+        /// <inheritdoc />
+        protected override int ItemCount => _itemCount;
+
+        /// <inheritdoc />
+        protected override void OnParametersSet()
+        {
+            if (Item == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires the template parameter '{nameof(Item)}' to be specified.");
+            }
+
+            if (Placeholder == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires the template parameter '{nameof(Placeholder)}' to be specified.");
+            }
+
+            if (ItemsProvider == null)
+            {
+                throw new InvalidOperationException($"{GetType()} requires a non-null parameter '{nameof(ItemsProvider)}'.");
+            }
+        }
+
+        /// <inheritdoc />
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            await base.OnAfterRenderAsync(firstRender);
+
+            if (_fetchState == FetchState.Waiting)
+            {
+                Debug.Assert(_fetchTask != null);
+
+                _fetchState = FetchState.Busy;
+
+                var result = await _fetchTask;
+
+                _itemCount = result.TotalItemCount;
+
+                foreach (var item in result.Items)
+                {
+                    _loadedItems.Enqueue(item);
+                }
+
+                _fetchTcs = null;
+                _fetchState = FetchState.Idle;
+
+                StateHasChanged();
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void RenderItems(RenderTreeBuilder builder)
+        {
+            var itemsRendered = 0;
+
+            foreach (var item in _loadedItems.Skip(ItemsBefore).Take(ItemsVisible))
+            {
+                Item(item)(builder);
```

Comment:
For the same reason as https://github.com/dotnet/aspnetcore/pull/24179/files#r461610727:

```suggestion
                builder.AddContent(0, Item, item);
```

... and the same thing when rendering `Placeholder` below.